File: src/main/java/com/gitlab/daring/image/command/transform/TransformCommands.java
package com.gitlab.daring.image.command.transform;

import com.gitlab.daring.image.command.Command;
import com.gitlab.daring.image.command.CommandRegistry;
import com.gitlab.daring.image.command.SimpleCommand;
import com.gitlab.daring.image.command.combine.BitwisePrevCommand;
import com.gitlab.daring.image.command.parameter.IntParam;

import static com.gitlab.daring.image.command.CommandUtils.newCommand;
import static com.gitlab.daring.image.util.ImageUtils.smat;
import static org.bytedeco.javacpp.opencv_core.*;
import static org.bytedeco.javacpp.opencv_imgproc.equalizeHist;

public class TransformCommands {

    public static void register(CommandRegistry r) {
        TransformCommands f = new TransformCommands();
        r.register("equalizeHist", f::equalizeHistCommand);
        r.register("bitwiseNot", f::bitwiseNotCommand);
        r.register("keepRange", f::keepRange);
        r.register("convert", ConvertCommand::new);
        r.register("morphology", MorphologyCommand::new);
        r.register("bitwisePrev", BitwisePrevCommand::new);
        r.register("normalize", NormalizeCommand::new);
        ThresholdCommands.register(r);
        FilterCommands.register(r);
        GeometricCommands.register(r);
    }

    public Command equalizeHistCommand(String... ps) {
        return newCommand(m -> equalizeHist(m, m));
    }

    public Command bitwiseNotCommand(String... ps) {
        return newCommand(m -> bitwise_not(m, m));
    }

    public Command keepRange(String... ps) {
        SimpleCommand c = new SimpleCommand(ps);
        IntParam lb = c.intParam(0, "0-255");
        IntParam ub = c.intParam(255, "0-255");
        return c.withFunc((m, d) -> {
            inRange(m, smat(lb.v), smat(ub.v), d); min(m, d, d);
        });
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/transform/TransformCommands.kt
package com.gitlab.daring.image.command.transform

import com.gitlab.daring.image.command.Command
import com.gitlab.daring.image.command.CommandRegistry

import com.gitlab.daring.image.command.CommandUtils.newCommand
import org.bytedeco.javacpp.opencv_core.bitwise_not
import org.bytedeco.javacpp.opencv_imgproc.equalizeHist

object TransformCommands {

    @JvmStatic
    fun register(r: CommandRegistry) {
        r.register("equalizeHist", this::equalizeHistCommand)
        r.register("bitwiseNot", this::bitwiseNotCommand)
        r.register("keepRange", ::KeepRangeCommand)
        r.register("convert", ::ConvertCommand)
        r.register("morphology", ::MorphologyCommand)
        r.register("normalize", ::NormalizeCommand)
        r.register("normalizeMean", ::NormalizeMeanCommand)
        r.register("applyClahe", ::CLAHECommand)
        ThresholdCommands.register(r)
        FilterCommands.register(r)
        GeometricCommands.register(r)
    }

    fun equalizeHistCommand(ps: Array<String>): Command {
        return newCommand { m -> equalizeHist(m, m) }
    }

    fun bitwiseNotCommand(ps: Array<String>): Command {
        return newCommand { m -> bitwise_not(m, m) }
    }

}
